libbpf-rs/src/query.rs

```diff
@@ -265,3 +266,105 @@ gen_info_impl!(
     libbpf_sys::bpf_btf_get_next_id,
     libbpf_sys::bpf_btf_get_fd_by_id
 );
+
+pub struct RawTracepointLinkInfo {
+    pub name: String,
+}
+
+pub struct TracingLinkInfo {
+    pub attach_type: ProgramAttachType,
+}
+
+pub struct CgroupLinkInfo {
+    pub cgroup_id: u64,
+    pub attach_type: ProgramAttachType,
+}
+
+pub struct NetNsLinkInfo {
+    pub ino: u32,
+    pub attach_type: ProgramAttachType,
+}
+
+pub enum LinkTypeInfo {
+    RawTracepoint(RawTracepointLinkInfo),
+    Tracing(TracingLinkInfo),
+    Cgroup(CgroupLinkInfo),
+    Iter,
+    NetNs(NetNsLinkInfo),
+    Unknown,
+}
+
+/// Information about a BPF link
+pub struct LinkInfo {
+    pub info: LinkTypeInfo,
+    pub id: u32,
+    pub prog_id: u32,
+}
+
+impl LinkInfo {
+    fn from_uapi(fd: i32, mut s: libbpf_sys::bpf_link_info) -> Option<Self> {
+        let type_info = match s.type_ {
+            libbpf_sys::BPF_LINK_TYPE_RAW_TRACEPOINT => {
+                if unsafe { s.__bindgen_anon_1.raw_tracepoint.tp_name } == 0 {
```

Comment:
you can be pretty sure that this is always going to be true. bpftool uses this check to get out of the infinite loop, but they way you implemented this there is no possibility of a loop, so you can slightly simplify this logic, just always specify tp_name and do obj_get_info() again to get details.